user/acl/man/man3/acl_from_mode.3
.\" Access Control Lists manual pages
.\"
.\" (C) 2002 Andreas Gruenbacher, <a.gruenbacher@computer.org>
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd March 23, 2002
.Dt ACL_FROM_MODE 3
.Os "Linux ACL"
.Sh NAME
.Nm acl_from_mode
.Nd create an ACL from file permission bits
.Sh LIBRARY
Linux Access Control Lists library (libacl, \-lacl).
.Sh SYNOPSIS
.In sys/types.h
.In acl/libacl.h
.Ft acl_t
.Fn acl_from_mode "mode_t mode"
.Sh DESCRIPTION
The
.Fn acl_from_mode
function creates a minimal ACL that contains the three entries with tag
types ACL_USER_OBJ, ACL_GROUP_OBJ, and ACL_OTHER, with permissions
corresponding to the owner, group, and other permission bits of its
argument
.Va mode .
.Sh RETURN VALUE
On success, this function returns a pointer to the
working storage. On error, a value of
.Li (acl_t)NULL
is returned, and
.Va errno
is set appropriately.
.Sh ERRORS
If any of the following conditions occur, the
.Fn acl_from_mode
function returns a value of
.Li (acl_t)NULL
and sets
.Va errno
to the corresponding value:
.Bl -tag -width Er
.It Bq Er ENOMEM
The ACL working storage requires more memory than is allowed by the
hardware or system-imposed memory management constraints.
.El
.Sh STANDARDS
This is a non-portable, Linux specific extension to the ACL manipulation
functions defined in IEEE Std 1003.1e draft 17 (\(lqPOSIX.1e\(rq, abandoned).
.Sh SEE ALSO
.Xr acl_equiv_mode 3 ,
.Xr acl_get_file 3 ,
.Xr acl 5
.Sh AUTHOR
Written by
.An "Andreas Gruenbacher" Aq a.gruenbacher@computer.org .


History
Ralink AP SDK - History

Version 3.4.1.0
===============

New Feature:

- [WebUI/NVRAM] add RT3092AP DLS GUI
- [WebUI] RVT GUI

App update:

- [Reset] adjust Default settings
- [Script] support GPL wifi driver insert/remove module

Fixed bug:

- [GPIO] Fix bug for GPIO=24~51
- [GPIOTOOL] add RALINK_GPIO_HAS_5124 for gpio24~gpio51
- [WebUI] STA site_survey error
- [WebUI] AP/STA display/Setting bugs
- [WebUI] WAPI/WPA_Supplicant Certificate upload
- [BusyBox] Sysklogd - avoid the read-only filesystem
- [Raether] MTU bug
- [Raether] Reorder the sequence of ei_close
- [Script] greenap.sh supports for MBSSID
- [Compile] sdk will compile goahead before mattrix, so we should modify goahead's Makefile to compile mattrix first
- [VPN] Multiple PPTP VPN pass through sessions are able to connect successfully for RT2880
- [NVRAM] rt3052 + i2s enabled => load default button's gpio=43

Version 3.4.0.0
===============

New Chip Support:

- [SDK] RT3883 support
- [SDK] RT3662 support
- [SDK] RT3350 support

New Feature:

- [Linux] Ralink Flow Classifier
- [Linux] Linux-based Watchdog driver (/dev/watchdog)
- [WebUI] WPS supplicant support
- [SDK] More 3G data card support
- [App] Video Flow Classification Command
- [App] Firewall-- Portscan, Syn flood  prevention. Snort Support.
- [App] iTune Server support
- [App] User space watchdog daemon
- [App] cURL
- [App] Expat- C library for parsing XML
- [App] mpstat
- [App] user space daemon supported 802.11f protocol
- [App] net-snmp 5.5
- [Web] Add WPS GUI for 2nd wireless interface
- [Web] Add WAPI GUI for 1st wireless interface
- [Web] Add WPA Supplicant GUI
- [Web] Add Wizard
- [Lib] Perl Compatible Regular Expressions
- [Lib] zlib-1.2.3
- [Lib] libpcap-1.0
- [Lib] libncurse-5.7

Peripheral Components:

- [I2S] audio codec WM8750 support for I2S TX/RX
- [PCM] Multi-SPI for PCM SLIC supported

Fixed bug:

- [PCM] assign next unmask channel every DMA done.
- [I2S] Change audio codec sampling rate to the nearest frequency settings.
- [WebUI] 3G PIN, APN, Password, Username, and Dial number support
- [WebUI] 3G auto-connect mode support
- [USB] OTG USB driver bug fix
- [NVRAM] fix commit length threshold
- [NVRAM] fix memory leak bug
- [App] IGMP Proxy issues fixed. (see History in user/igmpproxy/src/rt3052.c)


user/acl/libacl/acl_from_mode.c
/*
  File: acl_from_mode.c

  Copyright (C) 1999, 2000
  Andreas Gruenbacher, <a.gruenbacher@computer.org>

  This program is free software; you can redistribute it and/or
  modify it under the terms of the GNU Library General Public
  License as published by the Free Software Foundation; either
  version 2 of the License, or (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Library General Public License for more details.

  You should have received a copy of the GNU Library General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
*/

#include <unistd.h>
#include <stdio.h>
#include <errno.h>
#include <sys/stat.h>
#include <acl/libacl.h>
#include "libacl.h"


/*
  Create an ACL from a file mode.

  returns
  	the new ACL.
*/

acl_t
acl_from_mode(mode_t mode)
{
	acl_obj *acl_obj_p;
	acl_entry_obj *entry_obj_p;

	acl_obj_p = __acl_init_obj(3);
	if (!acl_obj_p)
		goto fail;

	entry_obj_p = __acl_create_entry_obj(acl_obj_p);
	if (!entry_obj_p)
		goto fail;
	entry_obj_p->etag = ACL_USER_OBJ;
	entry_obj_p->eid.qid = ACL_UNDEFINED_ID;
	entry_obj_p->eperm.sperm = (mode & S_IRWXU) >> 6;

	entry_obj_p = __acl_create_entry_obj(acl_obj_p);
	if (!entry_obj_p)
		goto fail;
	entry_obj_p->etag = ACL_GROUP_OBJ;
	entry_obj_p->eid.qid = ACL_UNDEFINED_ID;
	entry_obj_p->eperm.sperm = (mode & S_IRWXG) >> 3;

	entry_obj_p = __acl_create_entry_obj(acl_obj_p);
	if (!entry_obj_p)
		goto fail;
	entry_obj_p->etag = ACL_OTHER;
	entry_obj_p->eid.qid = ACL_UNDEFINED_ID;
	entry_obj_p->eperm.sperm = mode & S_IRWXO;
	return int2ext(acl_obj_p);

fail:
	__acl_free_acl_obj(acl_obj_p);
	return NULL;
}

